---
title: "TFM"
author: "Carmen Robinson Pastor"
date: "2023-04-17"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, warning=FALSE, message=FALSE}
#Download Bioconductor packages required for the execution of the code
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("TCGAbiolinks")
#BiocManager::install("DESeq2")
#BiocManager::install("ComplexHeatmap")
#BiocManager::install("org.Hs.eg.db")
#BiocManager::install("AnnotationDbi")
```

```{r, warning=FALSE, message=FALSE}
#Load packages needed for the execution of the code
library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(stringr)
library(tidyverse)
library(ggplot2)
library("org.Hs.eg.db")
library(AnnotationDbi)
library(RColorBrewer)
library(ggrepel)
library(survival) 
library(survminer)
library(glmnet)
library(dplyr)
library(devtools)
#devtools::install_github("imbs-hl/survivalsvm")
library(survivalsvm)
#library(rms)
#devtools::install_github("aroneklund/surviplot")
library(surviplot)
```

A1:

```{r}
#Obtain a summary of the chosen project
getProjectSummary("TCGA-LAML")

#Obtain data applying filters to the query to obtain only the data of interest
query <- GDCquery(
  project = "TCGA-LAML",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  access = "open",
  experimental.strategy = "RNA-Seq",
)

GDCdownload(
  query = query, 
  method = "api", 
  files.per.chunk = 6
)

data <- GDCprepare(query = query)

##Initial data exploration and preparation
#Get expression data
matrix <- assay(data) #this is our counts data
#dim(matrix) #we have 60660 genes of 151 samples
X<-as.data.frame(matrix)

X<- scale(t(X))
X<-X[, colSums(is.na(X))==0] #Remove columns with missing values

#get clinical data from the project
clinical_laml<- GDCquery_clinic("TCGA-LAML") 
```


A2:

```{r}
#extract gene and sample metadata from summarizedExperiment object
gene_metadata<- as.data.frame(rowData(data))
coldata <- as.data.frame(colData(data))

#censor alive patients out
#create status column for analysis
coldata$deceased<-ifelse(coldata$vital_status == "Alive", FALSE, TRUE) 

#create an "overall survival" variable that is equal to days_to_death for dead 
#patients, and to days_to_last_follow_up for patients that are still alive
coldata$overall_survival<- ifelse(coldata$vital_status == "Alive", 
                                        coldata$days_to_last_follow_up,
                                        coldata$days_to_death)

#Make sure the row names in sample_info match the column names in matrix
all(colnames(matrix) %in% rownames(coldata))
#Check if they are in the same order
all(colnames(matrix) == rownames(coldata))


#transform counts to be used in survival analysis
dds<-DESeqDataSetFromMatrix(countData= matrix, 
                            colData = coldata,
                            design = ~1)

#remove genes with sum total of 10 reads across all samples
keep<-rowSums(counts(dds))>= 10
dds<- dds[keep,]

#vst: variant stabilizing transformation
vsd<- vst(dds, blind=FALSE)
laml_matrix_vst<- assay(vsd)

#get data for a selection of genes and add gene metadata information to it
laml_sel<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id")  %>%
  filter(gene_name == "ELFN2" | gene_name == "LAMC3" | gene_name == "DLL3" | 
           gene_name == "PDPN"| gene_name == "GASK1A") 


#add clinical information to laml_sel
#we want to join by case IDs but the case IDs from our laml_sel and 
#clinical_laml are different, therefore we need to reformat the IDs
laml_sel$case_id <- gsub('-03.*', '', laml_sel$case_id)

laml_sel<- merge(laml_sel, clinical_laml, by.x = "case_id", 
                 by.y = "submitter_id")

#boxplot
ggplot(laml_sel, aes(x= laml_sel$gene_name, y = laml_sel$counts , 
                     color=laml_sel$vital_status)) +
  geom_boxplot() +
  ggtitle("Expression of a few genes") +
  xlab("Genes") +
  ylab("Expression") +
  theme(legend.title = element_blank())
```

A3: 

```{r}
## Construct a DESeqDataSet object
dds<- DESeqDataSetFromMatrix(countData = matrix, colData = coldata, 
                             design = ~vital_status)


#Pre-filtering of the data
#Remove rows with low gene counts: keep rows that have at least 10 reads 
keep<- rowSums(counts(dds)) >= 10 
#keep #all of them meet the criteria


#Set the factor level: set 'Alive' as our reference level
dds$vital_status <- relevel(dds$vital_status, ref= "Alive")

##Run DESeq
deseq <- DESeq(dds)

#Save the results
res<- results(deseq)

#Explore results
summary(res) #p-value cut off is 0.1 by default

#Change p-value cut off
res0.05<- results(deseq, alpha=0.05)
summary(res0.05) #1.1% are upregulated, 0.8% are downregulated
```

```{r}
res0.05.df<- as.data.frame(res0.05)

#Create a volcano plot using the differential gene expression analysis 
res0.05.df <- tibble::rownames_to_column(res0.05.df, "IDs")
res0.05.df$IDs <- gsub("\\.[0-9]*", "", res0.05.df$IDs) 
res0.05.df<- res0.05.df %>% remove_rownames %>% column_to_rownames(var= "IDs")

res0.05.df$IDs <- mapIds(org.Hs.eg.db, keys= rownames(res0.05.df), 
                         keytype= "ENSEMBL", column = "SYMBOL")


#add a column that indicates whether the genes are up or down regulated
res0.05.df$diffexpressed <- 'NO'
res0.05.df$diffexpressed[res0.05.df$log2FoldChange > 0.6 & 
                                res0.05.df$pvalue < 0.05] <- 'UP'
res0.05.df$diffexpressed[res0.05.df$log2FoldChange < -0.6 & 
                                res0.05.df$pvalue < 0.05] <- 'DOWN'

 
#find the top 20 differentially expressed genes
top20degs <- head(res0.05.df[order(res0.05.df$pvalue), 'IDs'], 20)
top20degs

#add a column
#if gene symbol (ID) is in the top 20 differentially expressed genes, assign it 
#the symbol,if not, assign NA
res0.05.df$delabel <- ifelse(res0.05.df$IDs %in% top20degs, res0.05.df$IDs, NA) 

#set a theme for the plot
#mirar un poco esto a ver si cambio alguna cosa
theme_set(theme_classic(base_size = 20) +
            theme(axis.title.y = element_text(face = "bold", 
                                              margin = margin(0,20,0,0), 
                                              size = rel(1.1), color = 'black'),
              axis.title.x = element_text(hjust = 0.5, face = "bold", 
                                          margin = margin(20,0,0,0), 
                                          size = rel(1.1), color = 'black'),
              plot.title = element_text(hjust = 0.5)
            ))


ggplot(data=res0.05.df, aes(x = log2FoldChange, y = -log10(pvalue),
                            col= diffexpressed, label= delabel)) + 
  geom_vline(xintercept = c(-0.6, 0.6), col= 'gray', linetype= 'dashed') + 
  geom_hline(yintercept = c(0.05), col= 'gray', linetype= 'dashed') +
  geom_point(size=2) +
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), 
                     labels = c("Downregulated", "Not significant", 
                                "Upregulated")) + 
  coord_cartesian(ylim=c(0, 20), xlim=c(-5, 5)) +
  scale_x_continuous(breaks = seq(-5, 5, 1)) + #breaks in the x axis
  labs(color= 'Diff. expressed genes', #legend title
       x=expression("log"[2]*"FoldChange"), y=expression("-log"[10]*"p-value"))+
  ggtitle("Gene expression in AML patients") + #plot title
  geom_text_repel(max.overlaps = Inf)
  
```

A4:

Lasso regression: 

```{r}
time<- coldata$overall_survival/30
time <- ifelse(time == "0", 1, time)
status<- coldata$deceased


fit <- Surv(time, status)
#the lengths of the fit object and the matrix need to have the same length
#find columns that have been omitted due to NA values
fit_m<- as.vector(fit)
which(is.na(fit_m))

fit <- na.omit(fit)


#Delete these rows from X
X2<- X[-c(8, 11, 22, 24, 44, 45, 72, 92, 103, 125, 145),]
dim(X2)


cv.lassoFit <- cv.glmnet(X2, fit, family="cox", alpha=1)
idealLambda <- cv.lassoFit$lambda.min
idealLambda1se <- cv.lassoFit$lambda.1se

co <- coef(cv.lassoFit, s=idealLambda)
rownames(co)[which(co!=0)]
co[which(co!=0)]

list<- rownames(co)[which(co!=0)]
list


list_g<- gsub("\\.[0-9]*", "", list)
list_g<- mapIds(org.Hs.eg.db, keys= list_g, keytype= "ENSEMBL", column="SYMBOL")
list_g

```

Kaplan Meier:

```{r}
#Combination of genes that improve survival when overexpressed
laml_over<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id") %>%
  filter(gene_name == "SERINC5" | gene_name == "RD3L" |
           gene_name == "RPS24P4" | gene_name == "LINC01258" | 
           gene_name == "LOC103171574" | gene_name == "MIR4285")

laml_over$case_id <- gsub('-03.*', '', laml_over$case_id)


#divide into high and low expression
#get median value
median_value_over<- median(laml_over$counts)

#which cases have higher or lower expression than median count
laml_over$strata <- ifelse(laml_over$counts >= median_value_over, "HIGH", "LOW")

laml_over<- merge(laml_over, clinical_laml, by.x = "case_id", by.y = "submitter_id")

over_fit <- survfit(Surv(overall_survival, deceased) ~ strata, data=laml_over)
over_fit

ggsurvplot(over_fit,
           data= laml_over,
           pval=T,
           risk.table=T)


over_fit2<- survdiff(Surv(overall_survival, deceased) ~ strata, data=laml_over)
over_fit2
```


```{r}
#Combination of genes that improve survival when underexpressed
laml_under<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id") %>%
  filter(gene_name == "RHOBTB2" | gene_name == "TREML2" | gene_name == "CBR1" | 
           gene_name == "SOCS1")


laml_under$case_id <- gsub('-03.*', '', laml_under$case_id)

#divide into high and low expression
#get median value
median_value_under<- median(laml_under$counts)

#which cases have higher or lower expression than median count
laml_under$strata <- ifelse(laml_under$counts >= median_value_under, "HIGH", "LOW")

laml_under<- merge(laml_under, clinical_laml, by.x = "case_id", by.y = "submitter_id")

under_fit <- survfit(Surv(overall_survival, deceased) ~ strata, data=laml_under)
under_fit

ggsurvplot(under_fit,
           data= laml_under,
           pval=T,
           risk.table=T)


under_fit2<- survdiff(Surv(overall_survival, deceased) ~ strata, data=laml_under)
under_fit2

```


Random forest:

```{r}
#laml<- laml_matrix_vst %>%
#  as.data.frame() %>%
#  rownames_to_column(var= "gene_id") %>%
#  gather(key = "case_id", value = "counts", -gene_id) %>%
#  left_join(.,gene_metadata, by="gene_id")  

#laml$case_id <- gsub('-03.*', '', laml$case_id)

#laml<- merge(laml, clinical_laml, by.x = "case_id", by.y = "submitter_id")

#X2 <- tibble::rownames_to_column(X2, "IDs")
#X2$IDs <- gsub("\\.[0-9]*", "", X2$IDs) 


#laml_surv<- laml[,c(74,75,76)]
#X2_surv<- merge(X2, laml)
 
#library(randomForestSRC)
#make train and test datasets, train is 70% of the data
#set.seed(120497)
#train<- sample(1:151, 106, replace=FALSE)
#train_laml<- laml[train,]
#test_laml<-laml[-train,]

#time_tr<- train_laml$overall_survival/30
#time_tr <- ifelse(time_tr == "0", 1, time_tr)
#status_tr<- train_laml$deceased
#rf_surv<- Surv(time_tr, status_tr)



#fit_rf<- rfsrc(rf_surv ~ ., data= as.data.frame(train_laml))
               #tree = 1000, importance=TRUE)
#fit_rf
#X2<- as.data.frame(X2)
#fit_try<- rfsrc(Surv(time, status) ~ ., data = X2, tree = 1000,
#                importance = TRUE)

```
