---
title: "TFM_final"
author: "Carmen Robinson Pastor"
date: "2023-06-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

DOWNLOAD AND LOAD PACKAGES:

```{r, warning=FALSE, message=FALSE}
#Download Bioconductor packages required for the execution of the code
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("TCGAbiolinks")
#BiocManager::install("DESeq2")
#BiocManager::install("ComplexHeatmap")
#BiocManager::install("org.Hs.eg.db")
#BiocManager::install("AnnotationDbi")
```

```{r, warning=FALSE, message=FALSE}
#Load packages needed for the execution of the code
library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(stringr)
library(tidyverse)
library(ggplot2)
library("org.Hs.eg.db")
library(AnnotationDbi)
library(RColorBrewer)
library(ggrepel)
library(survival) 
library(survminer)
library(glmnet)
library(dplyr)
library(devtools)
#devtools::install_github("imbs-hl/survivalsvm")
library(survivalsvm)
#devtools::install_github("aroneklund/surviplot")
library(surviplot)
library(Hmisc)
library(randomForestSRC)
```


OBTAINING AND PREPARING DATA:

```{r}
#Obtain a summary of the chosen project
getProjectSummary("TCGA-LAML")

#Obtain data applying filters to the query to obtain only the data of interest
query <- GDCquery(
  project = "TCGA-LAML",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  access = "open",
  experimental.strategy = "RNA-Seq",
)

GDCdownload(
  query = query, 
  method = "api", 
  files.per.chunk = 6
)

data <- GDCprepare(query = query)

##Initial data exploration and preparation
#Get expression data
matrix <- assay(data) 
dim(matrix) 
X<-as.data.frame(matrix)


#Get the sample data 
clinical_laml<- GDCquery_clinic("TCGA-LAML") #get clinical data from the project
#extract gene and sample metadata from summarizedExperiment object
gene_metadata<- as.data.frame(rowData(data))
coldata <- as.data.frame(colData(data))

#basic information about the data
#sex
table(coldata$gender)

#age
mean(coldata$age_at_index)
sd(coldata$age_at_index)

#race
table(coldata$race)

#vital status
table(coldata$vital_status)
```


```{r}
#censor alive patients out
#create status column for analysis
coldata$deceased<-ifelse(coldata$vital_status == "Alive", FALSE, TRUE) 

#create an "overall survival" variable that is equal to days_to_death for dead 
#patients, and to days_to_last_follow_up for patients that are still alive
coldata$overall_survival<- ifelse(coldata$vital_status == "Alive", 
                                        coldata$days_to_last_follow_up,
                                        coldata$days_to_death)

#Make sure the row names in sample_info match the column names in matrix
all(colnames(matrix) %in% rownames(coldata))
#Check if they are in the same order
all(colnames(matrix) == rownames(coldata))


#transform counts to be used in survival analysis
dds<-DESeqDataSetFromMatrix(countData= matrix, 
                            colData = coldata,
                            design = ~1)

#remove genes with sum total of 10 reads across all samples
keep<-rowSums(counts(dds))>= 10
dds<- dds[keep,]

#vst: variant stabilizing transformation
vsd<- vst(dds, blind=FALSE)
laml_matrix_vst<- assay(vsd)

#get data for a selection of genes and add gene metadata information to it
laml_sel<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id")  %>%
  filter(gene_name == "ELFN2" | gene_name == "LAMC3" | gene_name == "DLL3" | 
           gene_name == "PDPN"| gene_name == "GASK1A") 


#add clinical information to laml_sel
#we want to join by case IDs but the case IDs from our laml_sel and 
#clinical_laml are different, therefore we need to reformat the IDs
laml_sel$case_id <- gsub('-03.*', '', laml_sel$case_id)

laml_sel<- merge(laml_sel, clinical_laml, by.x = "case_id", 
                 by.y = "submitter_id")

#boxplot
ggplot(laml_sel, aes(x= laml_sel$gene_name, y = laml_sel$counts , 
                     color=laml_sel$vital_status)) +
  geom_boxplot() +
  ggtitle("Expression of a few genes") +
  xlab("Genes") +
  ylab("Expression") +
  theme(legend.title = element_blank())
```


DIFFERENTIAL GENE EXPRESSION BETWEEN ALIVE AND DEAD PATIENTS:

```{r}
## Construct a DESeqDataSet object
dds<- DESeqDataSetFromMatrix(countData = matrix, colData = coldata, 
                             design = ~vital_status)


#Pre-filtering of the data
#Remove rows with low gene counts: keep rows that have at least 10 reads 
keep<- rowSums(counts(dds)) >= 10 
#keep #all of them meet the criteria


#Set the factor level: set 'Alive' as our reference level
dds$vital_status <- relevel(dds$vital_status, ref= "Alive")

##Run DESeq
deseq <- DESeq(dds)

#Save the results
res<- results(deseq)

#Explore results
summary(res) #p-value cut off is 0.1 by default

#Change p-value cut off
res0.05<- results(deseq, alpha=0.05)
summary(res0.05) 
```

VOLCANO PLOT:

```{r}
res0.05.df<- as.data.frame(res0.05)

#Create a volcano plot using the differential gene expression analysis 
#Change gene ENSEMBL codes to gene symbols
res0.05.df <- tibble::rownames_to_column(res0.05.df, "IDs")
res0.05.df$IDs <- gsub("\\.[0-9]*", "", res0.05.df$IDs) 
res0.05.df<- res0.05.df %>% remove_rownames %>% column_to_rownames(var= "IDs")

res0.05.df$IDs <- mapIds(org.Hs.eg.db, keys= rownames(res0.05.df), keytype= "ENSEMBL", column = "SYMBOL")


#add a column that indicates whether the genes are up or down regulated
res0.05.df$diffexpressed <- 'NO'
res0.05.df$diffexpressed[res0.05.df$log2FoldChange > 0.6 & 
                                res0.05.df$pvalue < 0.05] <- 'UP'
res0.05.df$diffexpressed[res0.05.df$log2FoldChange < -0.6 & 
                                res0.05.df$pvalue < 0.05] <- 'DOWN'

 
#find the top 20 differentially expressed genes
top20degs <- head(res0.05.df[order(res0.05.df$pvalue), 'IDs'], 20)
top20degs

#add a column
#if gene symbol (ID) is in the top 20 differentially expressed genes, 
#assign it the symbol, if not, assign NA
res0.05.df$delabel <- ifelse(res0.05.df$IDs %in% top20degs, res0.05.df$IDs, NA) 

#set a theme for the plot
#mirar un poco esto a ver si cambio alguna cosa
theme_set(theme_classic(base_size = 20) +
            theme(axis.title.y = element_text(face = "bold", 
                                              margin = margin(0,20,0,0), 
                                              size = rel(1.1), color = 'black'),
              axis.title.x = element_text(hjust = 0.5, face = "bold", 
                                          margin = margin(20,0,0,0), 
                                          size = rel(1.1), color = 'black'),
              plot.title = element_text(hjust = 0.5)
            ))


ggplot(data=res0.05.df, aes(x = log2FoldChange, y = -log10(pvalue),
                            col= diffexpressed, label= delabel)) + 
  geom_vline(xintercept = c(-0.6, 0.6), col= 'gray', linetype= 'dashed') + 
  geom_hline(yintercept = c(0.05), col= 'gray', linetype= 'dashed') +
  geom_point(size=2) +
  scale_color_manual(values = c("#00AFBB", "grey", "#bb0c00"), #set the colours of our variable
                     labels = c("Downregulated", "Not significant", "Upregulated")) + #set labels
  coord_cartesian(ylim=c(0, 20), xlim=c(-5, 5)) +
  scale_x_continuous(breaks = seq(-5, 5, 1)) + #breaks in the x axis
  labs(color= 'Diff. expressed genes', #legend title
       x=expression("log"[2]*"FoldChange"), y=expression("-log"[10]*"p-value"))+
  ggtitle("Gene expression in AML patients") + #plot title
  geom_text_repel(max.overlaps = Inf)
  
```


SURVIVAL ANALYSIS:

DATA PREPARATION:

```{r}
#prerun
X<-as.data.frame(matrix)
coldata <- as.data.frame(colData(data))

coldata$deceased<-ifelse(coldata$vital_status == "Alive", FALSE, TRUE) 

coldata$overall_survival<- ifelse(coldata$vital_status == "Alive", 
                                        coldata$days_to_last_follow_up,
                                        coldata$days_to_death)
```


```{r}
#Shorten patient IDs
X<-t(X)
X<- as.data.frame(X)
X_m <- tibble::rownames_to_column(X, "patients")
X_m$patients <- gsub('-03.*', '', X_m$patients)
X_m<- X_m %>% remove_rownames %>% column_to_rownames(var= "patients")
X_m <- t(X_m)
#Gene ENSEMBL codes without extra numbers
X_m<- as.data.frame(X_m)
X_ma <- tibble::rownames_to_column(X_m, "IDs")
X_ma$IDs <- gsub("\\.[0-9]*", "", X_ma$IDs)
X_ma<- X_ma %>% remove_rownames %>% column_to_rownames(var= "IDs")
X_ma<- as.data.frame(X_ma)
#delete rows with NA values
#X_ma <- na.omit(X_ma)

# select individuals with expression data
coldata_s <- coldata[which(coldata$patient %in% colnames(X_ma)),]

#Make sure the patients in coldata match the rownames in X_ma
all(colnames(X_ma) %in% coldata_s$patient)

#Check if they are in the same order
all(colnames(X_ma) == coldata_s$patient)


#Make a list of ENSEMBL codes of the protein-coding genes
genes <- gene_metadata$gene_id[gene_metadata$gene_type == "protein_coding"]
genes <- gsub("\\.[0-9]*", "", genes)

gene_list <- genes


#survival analysis

time <- coldata_s$overall_survival/30
time <- ifelse(time == "0", 1, time)
status <- coldata_s$deceased

#initialize vectors
pval<- rep(NA, length=length(genes))
gene<- rep(NA, length=length(genes))

#loop on all genes from gene_list
for (i in c(1:length(gene_list)))
{
  #gene_list[i] = gene_list[i] + gene_list[i-1]
  #select gene
  gene <- as.character(gene_list[i])
  # If gene name in X_ma rownames, perform survival analysis, otherwise do nothing (as no data)
  if(gene %in% rownames(X_ma))
  {
    #select expression data for selected gene(s)
    e <- X_ma[rownames(X_ma)%in%gene , ]
    e <- t(e)
    #Compute the Tertiles and label them as low and high. Middle tertile is discarded (NA)
    clus <- ifelse(e<quantile(e,probs=1/3),"low",ifelse(e>quantile(e,probs=2/3),"high",NA)) 
 
  
    #compute survival model
    dfsurv <- data.frame(time, status, expr=clus)
    colnames(dfsurv)[3] <- "expr"
    try({
      fit <- survfit(Surv(time, status) ~ expr, data = dfsurv)
    # log-rank test p-value
      p1 <-surv_pvalue(fit)

       #plot KM curve
    png(file=paste("C:/Users/carme/Documents/Master_Bioinformatica/TFM/pngs/1gene", gene, ".png",sep=""))
    surviplot(Surv(time, status) ~ expr, data = dfsurv, col=c("blue", "red"), show.nrisk=F, hr.pos=NA, legend.pos = "topright", stitle=NA, cex=0.5)
    legend("top","LAML patients",bty="n",cex=0.9)
    legend(x=0, y=0.2, legend=as.character(p1$pval.txt), cex=0.8, box.lty = 0)
    legend(x=120, y=0.3,paste("Patients divided by tertiles", paste( "of ", gene, " expression.", sep=""), "Upper tertile compared to lower tertile", sep="\n"), cex=0.7, box.lty=0)
    dev.off()
      
      pval[i]<- p1$pval
    })
  }
  #print(i)
}

#save results in a table
res1 <- data.frame(gene, pval)

write.csv2(res1, file="C:/Users/carme/Documents/Master_Bioinformatica/TFM/Res.csv")

```

```{r}
#multiple testing correction
res1_mt<- res1
res1_mt$pval<- res1_mt$pval*19962

#Genes in gene column
genes <- gene_metadata$gene_id[gene_metadata$gene_type == "protein_coding"]
genes <- gsub("\\.[0-9]*", "", genes)

res1_mt$gene <- genes

res1_mt <- na.omit(res1_mt) 

#order by p-value and make the reduced dataset
res1_mt_red <- res1_mt[order(res1_mt$pval),]
res1_mt_red <- res1_mt_red[1:10,]

res1_mt_red$gene <- mapIds(org.Hs.eg.db, keys= res1_mt_red$gene, keytype= "ENSEMBL", column="SYMBOL")
```



```{r}
#Reduce dataset without multiple-testing correction
Res<- read.csv("C:/Users/carme/Documents/Master_Bioinformatica/TFM/Res.csv", 
               sep = ";", stringsAsFactors = TRUE)


#Genes in gene column
Res$gene <- genes


#order by p-value and make the reduced dataset
Res_red <- Res[order(Res$pval),]
Res_red <- Res_red[1:3082,]

#Make the reduced expression dataset
X_red <- X_ma[which(rownames(X_ma) %in% Res_red$gene),]
write.csv2(X_red, file="C:/Users/carme/Documents/Master_Bioinformatica/TFM/X_red.csv")

```


LASSO REGRESSION:

```{r}
# select individuals with expression data
coldata_red <- coldata_s[which(coldata_s$patient %in% colnames(X_red)),]


#Make sure the patients in coldata match the rownames in X_m
all(colnames(X_red) %in% coldata_red$patient)

#Check if they are in the same order
all(colnames(X_ma) == coldata_red$patient)

#transpose dataset
X_red <- t(X_red)

#time and status from reduced dataset
time_red<- coldata_red$overall_survival/30
time_red <- ifelse(time_red == "0", 1, time_red)
coldata_red$deceased <- as.numeric(coldata_red$deceased)
status_red<- coldata_red$deceased

#fit
fit <- Surv(time_red, status_red)
#delete missing values from data
which(is.na(fit))
fit <- na.omit(fit)
X_red<- X_red[-c(8, 11, 22, 24, 44, 45, 72, 92, 103, 125, 145),]


cv.lassoFit <- cv.glmnet(X_red, fit, family="cox", alpha=1)
idealLambda <- cv.lassoFit$lambda.min
idealLambda1se <- cv.lassoFit$lambda.1se

#obtain genes and their coefficients
co <- coef(cv.lassoFit, s=idealLambda)
rownames(co)[which(co!=0)]
co[which(co!=0)]
list<- rownames(co)[which(co!=0)]

list_g<- gsub("\\.[0-9]*", "", list)
list_g<- mapIds(org.Hs.eg.db, keys= list_g, keytype= "ENSEMBL", column="SYMBOL")
list_g

coefficients <- co[which(co!=0)]

lasso_res <- data.frame(list_g, coefficients)
lasso_res
```

KAPLAN MEIER:

```{r}
#Combination of genes that improve survival when overexpressed
laml_over<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id") %>%
  filter(gene_name == "RDL3" | gene_name == "GASK1A" |
           gene_name == "ZNF491" | gene_name == "BBIP1" | 
           gene_name == "LRRC37B" | gene_name == "SLC24A3" |
           gene_name == "TWIST1" )


laml_over$case_id <- gsub('-03.*', '', laml_over$case_id)

#divide into high and low expression
#get median value
median_value_over<- median(laml_over$counts)

#which cases have higher or lower expression than median count
laml_over$strata <- ifelse(laml_over$counts >= median_value_over, "HIGH", "LOW")

laml_over<- merge(laml_over, clinical_laml, by.x = "case_id", by.y = "submitter_id")

#Fit survival model
over_fit <- survfit(Surv(overall_survival, deceased) ~ strata, data=laml_over)
over_fit

#Plot
ggsurvplot(over_fit,
           data= laml_over,
           pval=T,
           risk.table=T)


over_fit2<- survdiff(Surv(overall_survival, deceased) ~ strata, data=laml_over)
over_fit2
```

```{r}
laml_under<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id") %>%
  filter(gene_name == "SNX8" | gene_name == "SFXN3" | gene_name == "ECE1" | 
           gene_name == "SEL1L3" | gene_name == "DAXX" | gene_name == "SDHA" |
           gene_name == "PPCDC" | gene_name == "BATF" | gene_name == "SOCS1") 


laml_under$case_id <- gsub('-03.*', '', laml_under$case_id)

#divide into high and low expression
#get median value
median_value_under<- median(laml_under$counts)

#which cases have higher or lower expression than median count
laml_under$strata <- ifelse(laml_under$counts >= median_value_under, "HIGH", "LOW")

laml_under<- merge(laml_under, clinical_laml, by.x = "case_id", by.y = "submitter_id")

under_fit <- survfit(Surv(overall_survival, deceased) ~ strata, data=laml_under)
under_fit

ggsurvplot(under_fit,
           data= laml_under,
           pval=T,
           risk.table=T)


under_fit2<- survdiff(Surv(overall_survival, deceased) ~ strata, data=laml_under)
under_fit2

```


RANDOM FOREST: 

```{r}

laml<- laml_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var= "gene_id") %>%
  gather(key = "case_id", value = "counts", -gene_id) %>%
  left_join(.,gene_metadata, by="gene_id")  

laml$case_id <- gsub('-03.*', '', laml$case_id)

laml<- merge(laml, clinical_laml, by.x = "case_id", by.y = "submitter_id")

X_red_n <- as.data.frame(X_red)


#only columns of interest for this part
laml_surv<- laml[,c(1,75,76)]
laml_surv <- laml_surv %>% distinct()
#write.csv2(laml_surv, file="C:/Users/carme/Documents/Master_Bioinformatica/TFM/laml_surv.csv")
#laml_surv <- read.csv("C:/Users/carme/Documents/Master_Bioinformatica/TFM/laml_surv.csv", sep = ";")

colnames(X_red_n) <- mapIds(org.Hs.eg.db, keys= colnames(X_red_n), keytype= "ENSEMBL", column="SYMBOL")

#delete column names that have the value NA
keep.cols <- names(X_red_n) %in% c(NA)
clean.X_red_n<- X_red_n[!keep.cols]

#merge datasets
clean.X_red_n <- tibble::rownames_to_column(clean.X_red_n, "IDs")

rf_surv<- merge(clean.X_red_n, laml_surv, by.x= "IDs", by.y= "case_id")
rf_surv<- rf_surv %>% remove_rownames %>% column_to_rownames(var= "IDs")
write.csv2(rf_surv, file="C:/Users/carme/Documents/Master_Bioinformatica/TFM/rf_surv.csv")


#make train and test datasets, train is 70% of the data
set.seed(1204)
train<- sample(1:151, 106, replace=FALSE)
train_laml<- rf_surv[train,]
test_laml<-rf_surv[-train,]

train_laml<-train_laml[complete.cases(train_laml),]#remove rows with NA values

train_laml$overall_survival<- train_laml$overall_survival/30
train_laml$overall_survival <- ifelse(train_laml$overall_survival == "0", 1, train_laml$overall_survival)
train_laml$deceased<- as.numeric(train_laml$deceased)

#fit the random survival forest model
fit_rf<- rfsrc(Surv(overall_survival, deceased) ~ . , data= train_laml, 
               ntree = 5, mtry=2, nsplit= 3, nodesize=30, seed= 1204,
               importance =TRUE)

fit_rf
png(file="C:/Users/carme/Documents/Master_Bioinformatica/TFM/fit_rf.png")
plot(fit_rf)
dev.off()

#C-index
cindex <- get.cindex(fit_rf$yvar[,1], fit_rf$yvar[,2], fit_rf$predicted.oob)
cindex

#prediction error
pe <- 1-cindex
pe
```

```{r}
#Permute importance
vimp<- vimp(fit_rf, importance= "permute")
vimp_importance <- data.frame(vimp$xvar.names, vimp$importance)

#variable selection 
varselect <- var.select(object= fit_rf, method= "vh.vimp")
varselect
```

```{r}
#predict on test data
pred <- predict(fit_rf, test_laml)
pred
```

